### ISSUE TITLE
XQuery 2

### ISSUE THREAD

**Issue by mmm202** [ebeshero/DHClass-Hub#163]
I'm stuck on number 4 where we need to use the "for loop" I just don't understand the "for loop" I guess and don't know where to put it or what variable to use

**Comment by jlm323** [comment]
I'm stuck on number 4 as well. I can't return the 707 distinct-values on `@ref`. This is what I've tried:

````
declare default element namespace "http://www.tei-c.org/ns/1.0";
let $letters := collection('/db/mitford')//body
let $contact := $letters//persName[distinct-values(string(@ref)="#id")]
return $contact
````

**Comment by RJP43** [comment]
@mmm202 Okay I am going to try and explain "for loop" as best I can so that hopefully you can see why we use it and how. Okay so we use for loop when we want the computer to parse through every SINGLE thing in a group of things. So in class Dr. B was showing how we took `for $play in $plays` so that the computer ran through EVERY play individually while giving them all the single variable of $play. If we didn't have the for loop you would have to create a variable for every single play file (or Mitford file in the case of XQuery 2) and then another variable for the conditions you want to place on every file. The for loop tells the computer I want you to process all of the files individually. An attempt at an analogy here: you go grocery shopping the whole lot of what you are buying could be called $groceries and each individual item as it is scanned through can be $grocery instead of having to mark each one ($banana, $bread, $alcohol (haha)) so then the computer knows oh okay you want everything inside of $groceries to be scanned under the same condition and that condition is determined on how you use $grocery later in your query). So now for how.... this sentence from the homework assignment sets up exactly want the single thing in the multiple of things you are looking for: `We want XQuery to look up for us which files are holding each entry on our list of distinct values` .... so you want each of the distinct `@refs` in the group of all of the distinct `@refs` from the entire collection so that you can compare them to the the `@refs` that appear in the file bodies (and that the part where Dr. B details the setting them equal to one another). Does this help?

**Comment by RJP43** [comment]
@jlm323 A couple issues here... I think you are trying to combine to many steps at once and if you break each part out into it's own part of the FLOWR statement it will be a lot clearer to you where you are going wrong. Right now since you have all of the info that you are actually trying to grab in a predicate (at this level) you are telling XQuery you want the placeNames that have all that... really though you want to take the distinct values of the `@ref` strings where they sit on the tree and then as i attempted to describe above you want to run through all those distinct refs individually (using for loop) to see when they appear in the files. Stretch this out a bit and try checking each variable as you go and I think you will get it. Let us know!

**Comment by RJP43** [comment]
Also I was able to use the actual `=` sign and instead had to use eq so that might be useful for everyone as they are completing this assignment. Dr. B has the comparison operators linked on the assignment but here is the link again: http://newtfire.org/dh/explainXPath.html#comparison

**Comment by mmm202** [comment]
I don't understand where the equal and greater than 15 are supposed to be and what I am supposed to have in my "for loop"

**Comment by nlottig94** [comment]
So I'm having trouble outputting what I want in a table. I am only getting two table rows instead of 11 (including the table headers). Help anyone?
Here's my XQuery:
`xquery version "3.0";
declare default element namespace "http://www.w3.org/1999/xhtml";
declare namespace tei="http://www.tei-c.org/ns/1.0";
<html>
    <head><title>Top Ten Most Referenced People in the Digital Mitford Project</title></head>
    <body>
        <table>
            <tr>
                <th>Person (@ref attribute)</th>
                <th>Files</th>
            </tr>
            <tr>
                <td>{
                let $mitford := collection('/db/mitford')
    			let $body := $mitford//tei:text/tei:body
    			let $person := $body//tei:persName/@ref/string()
    			let $distPers := distinct-values($person)
    			for $d in $distPers
    			let $ref := $body[.//tei:persName[@ref eq $d]]
    			where count($ref) gt 15
    			order by $d descending
    			return tokenize($d, '#')[last()]
                }</td>
                <td>{
                let $mitford := collection('/db/mitford')
    			let $body := $mitford//tei:text/tei:body
    			let $person := $body//tei:persName/@ref/string()
    			let $distPers := distinct-values($person)
    			for $d in $distPers
    			let $ref := $body[.//tei:persName[@ref eq $d]]
    			where count($ref) gt 15
    			return string-join((tokenize($mitford[$d]/base-uri(), '/')[last()]), ',')
                }</td>
            </tr>
        </table>
    </body>
</html>`

And here's my output XML:
`<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Top Ten Most Referenced People in the Digital Mitford Project</title>
</head>
<body>
<table>
<tr>
<th>Person (@ref attribute)</th>
<th>Files</th>
</tr>
<tr>
<td>Talfourd_Thos Shakespeare Russell_M Palmer_CF Mitford_Geo Macready_Wm MRM James_Miss Haydon Elford_SirWm</td>
<td>1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml 1821Nov12_13.Talfourd.xml</td>
</tr>
</table>
</body>
</html>`

**Comment by nlottig94** [comment]
I've just noticed that it's also not giving me the answers that it did before.  Also, when I tried to put onlt return statements in the `<td>` elements, and the rest of the FLWOR expression in the `<tr>` element _before_ the `<td>` it kept telling me that there was a curly brace where a closing tag should be...

**Comment by nlottig94** [comment]
I just tried this:
`xquery version "3.0";
declare default element namespace "http://www.w3.org/1999/xhtml";
declare namespace tei="http://www.tei-c.org/ns/1.0";
<html>
    <head><title>Top Ten Most Referenced People in the Digital Mitford Project</title></head>
    <body>
        <table>
            {
                let $mitford := collection('/db/mitford')
    			let $body := $mitford//tei:text/tei:body
    			let $person := $body//tei:persName/@ref/string()
    			let $distPers := distinct-values($person)
    			for $d in $distPers
    			let $ref := $body[.//tei:persName[@ref eq $d]]
    			where count($ref) gt 15
    			order by $d descending
                return
            }
            <tr>
                <th>Person (@ref attribute)</th>
                <th>Files</th>
            </tr>
            <tr>
                <td>{
    			tokenize($d, '#')[last()]
                }</td>
                <td>{
    			string-join((tokenize($mitford[$d]/base-uri(), '/')[last()]), ',')
                }</td>
            </tr>
        </table>
    </body>
</html>
`
And I'm still getting the same error...

**Comment by nlottig94** [comment]
I'm getting closer!!
This is now my XQuery:
`xquery version "3.0";
declare default element namespace "http://www.w3.org/1999/xhtml";
declare namespace tei="http://www.tei-c.org/ns/1.0";
<html>
    <head><title>Top Ten Most Referenced People in the Digital Mitford Project</title></head>
    <body>
        <table>
            <tr>
                <th>Person (@ref attribute)</th>
                <th>Files</th>
            </tr>
            {
                let $mitford := collection('/db/mitford')
    			let $body := $mitford//tei:text/tei:body
    			let $person := $body//tei:persName/@ref/string()
    			let $distPers := distinct-values($person)
    			for $d in $distPers
    			let $ref := $body[.//tei:persName[@ref eq $d]]
    			where count($ref) gt 15
    			order by $d descending
                return
            
            <tr>
                <td>{
    			tokenize($d, '#')[last()]
                }</td>
                <td>{
    			string-join((tokenize($mitford[$d]/base-uri(), '/')[last()]), ',')
                }</td>
            </tr>
            }
        </table>
    </body>
</html>
`
And this is my output:
`<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Top Ten Most Referenced People in the Digital Mitford Project</title>
</head>
<body>
<table>
<tr>
<th>Person (@ref attribute)</th>
<th>Files</th>
</tr>
<tr>
<td>Talfourd_Thos</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Shakespeare</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Russell_M</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Palmer_CF</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Mitford_Geo</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Macready_Wm</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>MRM</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>James_Miss</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Haydon</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
<tr>
<td>Elford_SirWm</td>
<td>1821Nov12_13.Talfourd.xml</td>
</tr>
</table>
</body>
</html>`
I now have all of the persNames, but only one file name...

**Comment by ebeshero** [comment]
@nlottig94 As you know, the problem is isolating the file names. Think about how you're doing it. What you have in your table cell is literally returning exactly what you're designating, which is the variable `$mitford` that happens to have `$d` (or one of the distinct values) in it. That variable was for the whole collection, and I think the base-uri() you're retrieving must be the very first one in the whole collection--not what you want.

The variable you want to define and return, for the filepath of the files that hold each $d, has to be written *inside* the for loop, because it's subject to the test of the for loop. Isn't your variable `$ref` already identifying the body of the file holding your `$d`? What would you need to do to `$ref` to return its file name?

**Comment by nlottig94** [comment]
Should I be tokenizing `$ref/base-uri()`?

**Comment by nlottig94** [comment]
I got it!

**Comment by brookestewart** [comment]
Okay, so I'm stuck on the first question. I'm guessing I'm missing something or it's a syntax error, but I can't figure it out. This is what I have 
`xquery version "3.0";
declare default element namespace "http://www.tei-c.org/ns/1.0";
base-uri(collection('/db/mitford')/*)` and I tried it without the `/*` and moved it around to different positions, but I keep getting an error. Can someone help please?

**Comment by mjb232** [comment]
I believe the Base-Uri function should go after your collection, just like a path step

**Comment by spadafour** [comment]
@brookestewart I hit the same issue. There are 2 folders directly under db/mitford, so when you say db/mitford/* , you are looking at two folders

**Comment by brookestewart** [comment]
Ah yes, I got it! Thank you guys so much!!!

**Comment by mjb232** [comment]
So Im stuck on the middle of four, right in the middle at the huge paragraph. I don't really know how to proceed. So far I have
`let $mit := collection('/db/mitford')`
`let $body := $mit//body`
`let $name := distinct-values($body//persName/@ref)`
`let $stmt := $mit//teiHeader//titleStmt`
`let $title :=$stmt/title/string()`
`for $n in $name`
`return $n`
Help!

**Comment by spadafour** [comment]
First off, your 2 variables let $stmt and let $title are doing nothing for yo; you can straight up ditch them. So what are you outputting right now? It should be the list of distinct values, except that list is repeated a lot of times. Thats because you right now are saying "for every distinct value in my list of distinct values, output my list of distinct values."

**Comment by mjb232** [comment]
`return $n` is giving me the 707 ref

**Comment by spadafour** [comment]
Ah, okay, right, hold on a sec...

**Comment by spadafour** [comment]
What we need to do is find a way to count those distinct values every time they show up in a body. So you're totally on the right track.

Let's walk this through in plain English before we try any FLWOR jargon. This is the part that stumped me, too.

We need a way to keep track of the amount of times a single @ref shows up in the body of a letter, and we need to be able to count it. This calls for a new variable. So we need to say that this new variable is equal to the bodies that have an @ref that's equal to any given @ref.

**Comment by mjb232** [comment]
so something like:
`let $bs := $body//[@ref eq $n]`  ?

**Comment by mjb232** [comment]
wait, no thats dumb. Give me a sec...

**Comment by spadafour** [comment]
Actually, you are pretty close

**Comment by mjb232** [comment]
`let $bs := $body[@ref eq $n]`  ?

**Comment by spadafour** [comment]
We're missing a key part: When you say $body[@whatever], you are looking at every single body, so we need a way to isolate/evaluate the exact place where the @ref is...

**Comment by spadafour** [comment]
Think about how we do this in XPath: we use the dot

**Comment by mjb232** [comment]
soooooooooo?
`let $bs := $body[.//@ref eq $n] `
I `return $bs` and got like 1000 results, is that right?

**Comment by spadafour** [comment]
You almost have it (I'm pretty sure that line is correct), but we don't want to return $bs. What do we actually want to return? We want to return @ref (in your case $n) where the total count of the bodies $n appears in is greater than 15

**Comment by mjb232** [comment]
So I need a `where` to filter my return, and then im actually returning `$n`, soooo like 
`let $bs := $body[.//@ref eq $n]`
`where count ($bs) gt 15`
`return $n`

**Comment by spadafour** [comment]
I think that should work; what are you getting? (It takes a little bit to process)

**Comment by mjb232** [comment]
still ten items and it finished super quick.....

**Comment by spadafour** [comment]
Ten items is right... what are the @refs?

**Comment by mjb232** [comment]
#Talfourd_Thos
#Macready_Wm
#Mitford_Geo
#MRM
#Russell_M
#James_Miss
#Shakespeare
#Palmer_CF
#Elford_SirWm
#Haydon

**Comment by spadafour** [comment]
Yep, thats it

**Comment by mjb232** [comment]
awesome! You helped sooooooooooo much!

**Comment by laurenmcguigan** [comment]
So for number 7 on the bonus This is my code:

````
xquery version "3.0";
declare default element namespace "http://www.w3.org/1999/xhtml";
declare namespace tei="http://www.tei-c.org/ns/1.0";
  <html>
<head><title>Top Ten Most Referenced People in the Digital Mitford Project</title></head>
<body>
    <table>
        <tr><th></th><th></th></tr>
{
let $collection:= collection('/db/mitford')
let $body:= $collection//tei:body
let $persName:= $body//tei:persName
let $ref:= $persName/@ref
let $distinctref:= distinct-values($ref)
for $d in $distinctref
let $fileref:= $body[.//tei:persName[@ref eq $d]]
let $fileNames:= $fileref/base-uri()
let $tokenref:= tokenize($d,"#")[last()]
where count($fileref) gt 15
order by $tokenref descending
return 
    <tr><td>
        {$tokenref}
        </td>
        </tr>
        
}
{
    let $collection:= collection('/db/mitford')
let $body:= $collection//tei:body
let $persName:= $body//tei:persName
let $ref:= $persName/@ref
let $distinctref:= distinct-values($ref)
for $d in $distinctref
let $fileref:= $body[.//tei:persName[@ref eq $d]]
let $fileNames:= $fileref/base-uri()
for $f in $fileNames
let $tokenref:= tokenize($d,"#")[last()]
where count($fileref) gt 15
order by $tokenref descending
return 
         <tr>
        <td>
            {$f}
            </td>
            </tr>
            
       
}

</table>
</body>
</html>
````

But I'm getting the names separate from files. I believe it is just an issue with how the tables are set up, but I'm not sure..

**Comment by laurenmcguigan** [comment]
Sorry I accidentally closed it.  Can you see my comment?

**Comment by ebeshero** [comment]
@laurenmcguigan I didn't run your code, but just staring at it, I think what's happening is you've wrapped the names and the results inside two different table rows or `<tr>` . You want the two to be sitting side-by-side in the same row. And, it's a little more complicated, what you need to do with the HTML. Here's how the whole table structure should come out. There's actually a **whole table** sitting inside a table cell!

````
<table>
<tr>
    <td>{NAME}</td>
     <td>
            <table>
                   <tr> 
                        <td>{File}</td>
                   </tr>
              </table>
        </td>
     </tr>
</table>

**Comment by ebeshero** [comment]
So, imagine what's happening is, a whole table of files is sitting inside the cell next door to the cell with the person's name!

**Comment by laurenmcguigan** [comment]
@ebeshero  so I did this:
````
xquery version "3.0";
declare default element namespace "http://www.w3.org/1999/xhtml";
declare namespace tei="http://www.tei-c.org/ns/1.0";
  <html>
<head><title>Top Ten Most Referenced People in the Digital Mitford Project</title></head>
<body>
    <table>
        <tr><th></th><th></th></tr>
{
let $collection:= collection('/db/mitford')
let $body:= $collection//tei:body
let $persName:= $body//tei:persName
let $ref:= $persName/@ref
let $distinctref:= distinct-values($ref)
for $d in $distinctref
let $fileref:= $body[.//tei:persName[@ref eq $d]]
let $fileNames:= $fileref/base-uri()
for $f in $fileNames
let $tokenref:= tokenize($d,"#")[last()]
where count($fileref) gt 15
order by $tokenref descending
return 
    <table>
        <tr>
            <td>
                {$tokenref}
            </td>
        </tr>
         <tr>
            <td>
                 {$f}
            </td>
        </tr>
    </table>    
}

</table>
</body>
</html>
````
But now I'm getting the name above EACH file

**Comment by ebeshero** [comment]
@laurenmcguigan Right--and that makes total sense to the computer reading your code! ;-) But it's not what you want. So you have that OUTER table all set up in the outer HTML structure. You started it before you opened your FLWOR. So when you output your return, you're setting up each table row inside, and the table cells inside that.

Inside your second table cell (`<td>`), you want to position your new table! That's how you get a table living inside a table cell.

**Comment by laurenmcguigan** [comment]
@ebeshero 
````
xquery version "3.0";
declare default element namespace "http://www.w3.org/1999/xhtml";
declare namespace tei="http://www.tei-c.org/ns/1.0";
  <html>
<head><title>Top Ten Most Referenced People in the Digital Mitford Project</title></head>
<body>
    <table>
        <tr><th></th><th></th></tr>
{
let $collection:= collection('/db/mitford')
let $body:= $collection//tei:body
let $persName:= $body//tei:persName
let $ref:= $persName/@ref
let $distinctref:= distinct-values($ref)
for $d in $distinctref
let $fileref:= $body[.//tei:persName[@ref eq $d]]
let $fileNames:= $fileref/base-uri()
for $f in $fileNames
let $tokenref:= tokenize($d,"#")[last()]
where count($fileref) gt 15
order by $tokenref descending
return 
    <table>
        <tr> 
            <td>{$tokenref} </td>
        </tr>
         <tr>
        <td>
            <table>
            {$f}
            </table>
            </td>
            </tr>
    </table>        
            
       
}

</table>
</body>
</html>
````
I'm still getting the names before each file??

**Comment by ebeshero** [comment]
@laurenmcguigan A couple of problems:
1) You have three table levels here, but you only need two! You need the one outer table, and then just the table inside the cell.
2) Okay. This is kind of the bonus part of the challenge. You know how you output one name in each cell in one column? You want the cell right next to it to contain a whole table...and you've *started* to set that up here...but think about numbers. I mean, for *every single $tokenref* you have a *whole list of files*. So think about where you are positioning that second "for loop." You are right that you need it, but you need to position it *after* you've output the simple column that contains the 10 names. So, set your table, and realize that every table needs to have table rows and table cells as always, and inside each one of those *inner* table cells you get your output of `$f`. It has to be output as part of an inner "for loop" that runs every time you output a name.

**Comment by ebeshero** [comment]
@laurenmcguigan So...the reason you are currently getting a name output over and over again *for every file* is because you've set up that second "for loop" a little too soon. Literally, the computer is giving back what you asked. You said: Computer! Every time you see a file name, output the `$tokenref` associated with it. In reality, for this table, you only want `$tokenref` output 10 times and not be repeated. You don't want `$tokenref` to be output inside that inner for loop.

**Comment by laurenmcguigan** [comment]
@ebeshero I just don't understand how to return only one of the names in the table and not one for each

**Comment by laurenmcguigan** [comment]
@ebeshero nevermind, you already explained it, I just don't understand it. Maybe we could discuss it in class tomorrow?

**Comment by ebeshero** [comment]
Yep! Meanwhile, you all have a nice new Decameron file in beautiful TEI with floatingText that I am excited to show you all. I wanted to bootstrap that for the team to help you study the text...
@jlm323 will need to review my pull request.
